File: java/src/main/java/stack/DailyTemperatures.java
package stack;

import java.util.List;
import java.util.Stack;

class Tuple {

    public Tuple(int temp, int index) {
        this.temp = temp;
        this.index = index;
    }
    int temp;
    int index;
}

public class DailyTemperatures {

    // O(N) solution.
    public static int[] dailyTemperatures(int[] temps) {
        int[] toRet = new int[temps.length];
        Stack<Tuple> stack = new Stack<>();

        // Idea: We want the stack to be in 'monotonic decreasing order'. So iterate through
        // the array and add to the stack in decreasing order. If we encounter something greater than our min, then pop off the stack
        // until we reach a new min or the end of the stack.
        for (int i = 0; i < temps.length; i++) {

            // Stack.peek() should always be the min value we have seen thus far.
            while (!stack.isEmpty() && temps[i] > stack.peek().temp) {

                Tuple pop = stack.pop();
                // compare the index of what was on the stack to what we are currently at.
                // Set the days since by subtracting current day to the day the temp was pushed on (pop.index)

                toRet[pop.index] = i - pop.index;
            }
            stack.push(new Tuple(temps[i], i));

        }

        return toRet;
    }

    public static void main(String[] args) {

        int[] in = {22,21,24};

        int[] out = dailyTemperatures(in);

        System.out.printf("Input: %s\n Output: %s", List.of(in).toString(), List.of(out).toString());

    }
}

File: cpp/stack/dailyTemperatures.cpp
#include <vector>
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
      
      vector<int> toRet(temperatures.size());
      stack<pair<int, int>> theStack;


      for (int i = 0; i < temperatures.size(); i++) {
        
        if (theStack.empty()) {
          theStack.push({temperatures[i], i});
        } else {


          // Pop off pairs that are smaller than the current temperature.
          while (!theStack.empty() && temperatures[i] > theStack.top().first) {
            // populate the result array.
            
            toRet[theStack.top().second] = i - theStack.top().second; // Days since that temperature is i - p.second (index we pushed onto the stack.) 
            theStack.pop(); // days have been found, pop that bitch off
          }

          theStack.push({temperatures[i], i});
        }
      }



      return toRet;
    }
};



int main() {


  vector<int> input = {30, 38, 30, 36, 35, 40, 28};
  Solution s;

  cout << "For input: [";
  copy(input.begin(), input.end(), ostream_iterator<int>(cout, " "));
  cout << "]" << endl;

  vector<int> res = s.dailyTemperatures(input);
  cout << "Result: [";
  copy(res.begin(), res.end(), ostream_iterator<int>(cout, " "));
  cout << "]" << endl;


  return 0;
}
